internal/harness/evaluator/golang/apisrv/apisrv.go

```diff
@@ -0,0 +1,296 @@
+				start = time.Now()
+				time.Sleep(time.Duration(pollInterval) * time.Millisecond)
+			}
+
+			continue
+		}
+
+		numRunning, err := strconv.Atoi(r)
+		if err != nil {
+			logger.Errorf("Failed to retrieve number of currently running MMFs, %v", err)
+		}
+
+		// Some MMFs ran since we last evaluated. We should evaluate the proposal either when all
+		// MMFs are complete, or the timeout is reached.
+		switch {
+		// dividing by 1e6 looks gnarly but is actually the canonical implementation
+		// https://github.com/golang/go/issues/5491
+		case time.Since(start).Nanoseconds()/1e6 >= maxWait:
+			logger.Infof("Maximum evaluator wait time (%v ms) exceeded", maxWait)
+			ctx, _ = tag.New(ctx, tag.Insert(KeyEvalTrigger, "interval_exceeded"))
+			checkProposals = true
+
+		case numRunning <= 0: // Can be less than zero in some less common cases.
+			logger.Info("All MMFs complete, trigger evaluation")
+			ctx, _ = tag.New(ctx, tag.Insert(KeyEvalTrigger, "mmfs_completed"))
+			numRunning = 0
+			checkProposals = true
+		}
+
+		if checkProposals {
+			// Only run the evaluator if there are queued proposals.
+			checkProposals = false
+			results, err := redishelpers.Count(context.Background(), s.Pool, s.Config.GetString("queues.proposals.name"))
+			switch {
+			case err != nil:
+				logger.Errorf("Failed to retrieve the length of the proposal queue from statestorage, %v", err)
+			case results == 0:
+				logger.Info("no proposals in the queue to evaluate")
+			default:
+				logger.Infof("%v proposals available, evaluating", results)
+				go s.evaluator(ctx)
+			}
+
+			// Delete the counter.  This causes subsequent queries to return 'nil' until it is
+			// re-created again on first write.  This enables us to skip evaluation logic if no MMF
+			// has written to the counter since last evaluation.
+			err = redishelpers.Delete(context.Background(), s.Pool, "concurrentMMFs")
+			if err != nil {
+				logger.WithFields(log.Fields{
+					"error": err.Error(),
+				}).Error("error deleting concurrent MMF counter")
+			}
+			start = time.Now()
+		}
+
+		// A sleep here is not critical but just a useful safety valve in case
+		// things are broken, to keep the main loop from going all-out and spamming the log.
+		logger.WithFields(log.Fields{"elapsed ": time.Since(cycleStart)}).Info("Evaluation complete. Sleeping for 1 sec.")
+		time.Sleep(time.Duration(1000) * time.Millisecond)
+	}
+}
+
+// evaluator queries all the proposals from Open Match and calls the user's evaluator with this collection
+// of proposals. It receives approved proposals which it then updates in Open Match as results.
+func (s *Evaluator) evaluator(ctx context.Context) error {
+	logger := s.Logger
+	logger.Info("Starting evaluation")
+
+	// Get connection to redis
+	redisConn := s.Pool.Get()
+	defer redisConn.Close()
+
+	start := time.Now()
+	// Get the number of proposals currently queued for evaluation.
+	proposalq := s.Config.GetString("queues.proposals.name")
```

Comment:
Not obvious that this is a name of the proposal queue. queueName?

Reply:
Done!